src/lib/y2configuration_management/clients/formula.rb

```diff
@@ -0,0 +1,139 @@
+require "yast"
+require "y2configuration_management/salt/formula"
+require "y2configuration_management/salt/form_controller"
+require "y2configuration_management/salt/formula_selection"
+
+module Y2ConfigurationManagement
+  module Clients
+    # Client to configure formulas
+    class Formula < Yast::Client
+      include Yast::Logger
+      extend Yast::I18n
+
+      attr_accessor :formulas
+      attr_reader :states_root, :formulas_root, :pillar_root
+
+      def main
+        textdomain "configuration_management"
+        import_modules
+        configure_directories
+        read_formulas
+
+        # Mechanism to detect if we're going back
+        @last_formula_idx = 0
+
+        start_workflow
+      end
+
+    private
+
+      # This code is still experimental, so let's disable this check.
+      # rubocop:disable Metrics/AbcSize
+      def start_workflow
+        sequence = {
+          "ws_start"        => "choose_formulas",
+          "choose_formulas" => {
+            abort: :abort,
+            next:  formulas.empty? ? :next : formulas[0].name
+          },
+          "apply_formulas"  => {
+            abort: :abort,
+            next:  :next
+          }
+        }
+
+        workflow_aliases = {
+          "choose_formulas" => ->() { choose_formulas },
+          "apply_formulas"  => ->() { apply_formulas }
+        }
+
+        formulas.each_with_index do |formula, idx|
+          sequence[formula.name] = {
+            abort:  :abort,
+            cancel: "choose_formulas",
+            next:   idx < formulas.size - 1 ? formulas[idx + 1].name : "apply_formulas",
+            back:   idx > 0 ? formulas[idx - 1].name : "choose_formulas"
+          }
+          workflow_aliases[formula.name] = ->() { parametrize_formula(formula) }
+        end
+
+        log.info "Starting formula sequence"
+        log.info "Aliases: #{workflow_aliases.inspect}"
+        log.info "Sequence: #{sequence.inspect}"
+
+        Sequencer.Run(workflow_aliases, sequence)
+      end
+
+      # This code is still experimental, so let's disable this check.
+      def choose_formulas
+        unless formulas && !formulas.empty?
```

Comment:
```suggestion
        if Array(formulas).empty?
```
this normalizes `nil` to `[]`